src/transformers/models/llama/modeling_llama.py

```diff
@@ -378,16 +387,18 @@ def forward(
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
-            kv_seq_len += past_key_value[0].shape[-2]
+            if self.layer_idx is None:
+                raise ValueError(
+                    f"The cache structure has changed since version v4.36. If you are using {self.__class__.__name__} "
+                    "for auto-regressive decoding with k/v caching, please make sure to initialize the attention class "
+                    "with a layer index."
+                )
+            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
         if past_key_value is not None:
-            # reuse k, v, self_attention
-            key_states = torch.cat([past_key_value[0], key_states], dim=2)
-            value_states = torch.cat([past_key_value[1], value_states], dim=2)
-
-        past_key_value = (key_states, value_states) if use_cache else None
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cos, sin)
```

Comment:
We will very soon implement a static k/v cache here (as shown in https://github.com/pytorch-labs/gpt-fast). To implement a k/v cache, we need to mask future (not-yet-attended) attention mask tokens so that they don't influence the result. 

This means we also need to pass the `attention_mask` to the `update(...)` function. We could solve this by making `update(...)` accept a dict instead of just tensors: https://github.com/huggingface/transformers/pull/26681/files#r1405895192